### PR TITLE
Define metal::accumulate for zero input lists

### PR DESCRIPTION
**Pull request by brunocodutra**
[Speaking of accumulate](https://github.com/brunocodutra/metal/pull/100#issuecomment-542378872) I decided to have a look at it and realized it could be simplified if implemented in terms of `metal::transform`.

### CODE REVIEW

**Comment by brunocodutra** [comment]
@ecrypa would love your feedback on this.

**Comment by brunocodutra** [comment]
[sigh](https://ci.appveyor.com/project/brunocodutra/metal/builds/28164176/job/5qfn34q3myn8k98b#L113).

File: include/metal/list/accumulate.hpp [hunk 1]
```diff
@@ -46,54 +46,44 @@ namespace metal {
...
-
-        template<class lbd>
-        struct _accumulate {};
-
-        template<template<class...> class expr>
-        struct _accumulate<lambda<expr>> {
-            template<class state, class... seqs>
-            using type =
-                forward<_accumulate_impl<state, seqs...>::template type, expr>;
+        struct _accumulate<state, list<vals...>> {
```
[the remark this answers is not in the document]
Comment (brunocodutra) [reply include/metal/list/accumulate.hpp]: Yes, by far the most common use case I expect to be accumulating a single list.

>  I wonder if we can also increase performance for a total number of two, three, and, say, four seqs.

We don't need to, that's the beauty of this version! Notice that it relies on `transpose`, which already specializes on 1, 2 and 3 lists. That's probably why `accumulate` got faster in your benchmark.

**Comment by brunocodutra** [comment]
> With Clang 9 and GCC 9 it is also faster than the old accumulate version of min_element.

Wow nice, didn't expect that to be honest.

> I did see the troubles of old compilers.

I hate MSVC with passion.

File: include/metal/list/accumulate.hpp [hunk 2]
```diff
@@ -46,54 +46,44 @@ namespace metal {
...
+        if_<same<size<seqs>...>,
+            detail::_accumulate<state, seqs...>>::template type,
+        lbd>;
 }
 
+#include "../lambda/apply.hpp"
 #include "../lambda/lambda.hpp"
+#include "../lambda/partial.hpp"
 #include "../list/at.hpp"
 #include "../list/indices.hpp"
```
[the remark this answers is not in the document]
Comment (brunocodutra) [reply include/metal/list/accumulate.hpp]: Absolutely, thanks for catching this one!

**Comment by brunocodutra** [comment]
While playing with this I realized `metal::accumulate` could be generalized to accept zero input lists and decided to change this PR to that instead.

**Comment by ecrypa** [comment]
Beautiful!

Besides, the discarded `min_element` index retrieval is even faster with this version (v21). See [Clang 9](https://ecrypa.github.io/metal-benchmark/min_element_clang9_v4_v21_accumulate3/) and [GCC 9](https://ecrypa.github.io/metal-benchmark/min_element_gcc9_v4_v21_accumulate3/).

**Comment by brunocodutra** [comment]
Woah nice, thanks for checking!

**Pull request merged**